### ISSUE TITLE
No errors given for invalid named imports from package in node_modules

### ISSUE THREAD

**Issue by jamietre** [benmosher/eslint-plugin-import#765]
This is most likely a support request,  but this just doesn't seem to work for me and I've tried a variety of configs. I can't get errors to be reported for invalid imports from `node_modules` dependencies. Errors are reported as expected from modules within the application itself, and I've also been able to get `eslint-import-resolver-babel-module` to correctly resolve modules with local aliases. But I can't get it to work with things in `node_modules`:

`
import { somethingNotExported } from 'some_module'
`

Tried basic config:

    "extends":  "plugin:import/errors"

Tried explicitly specifying node resolution:

    "settings": {
        "import/resolver": {
            "node": {}
        }
    },

Tried explicitly setting rule:

        "import/no-unresolved": [2, {"commonjs": true }]

Any ideas what I'm doing wrong? No errors are ever reported for imports from `node_modules`, except for the actual package not existing.

-- edit -

This may be me simply not understanding the capabilities here. Of course the exports of packages in node_modules are not in ES6 syntax for the most part.  Buit is there a possible configuration that can parse typescript typings? In the case of modules where I'm trying to get syntax validation, there are typings present.

**Comment by DanielSWolf** [comment]
I'm having the same problem. In my case, the other package actually *uses ES6 exports*. It's a private NPM package that we publish as ES2015. Still, I'm not getting an error if I import a non-existing named export.

**Comment by DanielSWolf** [comment]
I solved at least part of the problem. There is an ESLint setting `"import/ignore"` that contains an array of regex strings. Any file whose path matches one of those regexes will be ignored. This setting has the following default value:

```js
"import/ignore": ["node_modules", "\\.(coffee|scss|css|less|hbs|svg|json)$"]
```

That means that any file in `node_modules` will be ignored.

Explicitly setting the value to `["\\.(coffee|scss|css|less|hbs|svg|json)$"]` will fix the problem *at least for ES6 exports*. For `module.export`-style exports, this still doesn't help, though.

**Comment by ljharb** [comment]
Since npm-installed modules should always be transpiled to CJS, this generally shouldn't be an issue.

**Comment by jay1337** [comment]
@ljharb it's the opposite: ES6 exports are working but CJS modules doen't work ! I have the same issue than @DanielSWolf .

**Comment by ljharb** [comment]
@jay1337 can you provide the exact import line that should be failing, and your eslintrc?

**Comment by jay1337** [comment]
@ljharb here is what you asked for:

```
import { anything-not-existing-in-existing-valid-installed-module } from 'existing-valid-installed-module`;
```

```
{
	"extends": [
		"eslint:all",
		"plugin:import/errors",
		"plugin:import/warnings",
	],
	"parserOptions": {
		"ecmaVersion": 6,
		"sourceType": "module",
		"allowImportExportEverywhere": true,
	},
	"env": {
		"es6": true,
		"node": true,
	},
	"parser": "babel-eslint",
	"plugins": [
		"import",
	],
	"settings": {
		"import/resolver": "node",
		"import/ignore": [],
	},
}
```

**Comment by ljharb** [comment]
@jay1337 Thanks! can you provide a specific module and export that doesn't work? Also, why do you need the `allowImportExportEverywhere` setting?

**Comment by jay1337** [comment]
@ljharb There is an issue with many modules, you can test with `import { blabla } from 'express';`

I use `allowImportExportEverywhere` for conditional imports. It is useful with Meteor to import files on server but not on client (in a file shared between server and client).

**Comment by ljharb** [comment]
Rereading https://github.com/benmosher/eslint-plugin-import/blob/master/docs/rules/namespace.md, it looks like this just isn't a feature this plugin supports. "A module path that is ignored or not unambiguously an ES module will not be reported when imported."

Node itself doesn't know how to map named imports to CJS `module.exports`, which is why the current plan for node is to not allow anything but default imports from CJS modules.

**Comment by jay1337** [comment]
@ljharb you are right ! However there is a new way for modules to give the path of their ES source. They now use `module` field in `package.json` (instead of `jsnext:main`). It's gonna be the new standard : https://github.com/dherman/defense-of-dot-js/blob/master/proposal.md#typical-usage . Some important modules have already switched to `module`field : `react-router`, ... Could `eslint-plugin-import` check also this field ?

**Comment by ljharb** [comment]
It’s absolutely not going to be a “new standard”; that link is to a failed proposal. The module field is non-standard, and primarily because people are already using it, it will likely never be (it’d have to pick a new name).

**Comment by jay1337** [comment]
Ok. Up to you. But looking at all my dependencies, a lot are using module
field. I think it would be very useful. If it's not a standard you could
add it as an option (switch to enable, or list of field names, or regex,
...).

**Comment by ljharb** [comment]
We could; but I'd personally oppose it. Spreading and enabling the use of nonstandard package.json fields is bad for the ecosystem.
